File: backend/src/main/java/com/trading/controller/MarketController.java
package com.trading.controller;

import com.trading.service.MarketService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/market")
public class MarketController {

    private final MarketService marketService;

    public MarketController(MarketService marketService) {
        this.marketService = marketService;
    }

    @GetMapping("/{symbol}")
    public ResponseEntity<MarketService.PriceData> getMarketData(
            @PathVariable String symbol, @RequestParam(defaultValue = "30") int days) {
        return ResponseEntity.ok(marketService.getSharePrice(symbol));
    }
}

File: src/main/kotlin/com/trading/controller/MarketController.kt
package com.trading.controller

import com.trading.service.MarketService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/mcp/market")
class MarketController(
    private val marketService: MarketService
) {
    
    @PostMapping("/tools/lookup_share_price")
    fun lookupSharePrice(@RequestBody request: LookupSharePriceRequest): ResponseEntity<ToolResponse<Double>> {
        return try {
            val price = marketService.getSharePrice(request.symbol)
            ResponseEntity.ok(ToolResponse.success(price))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(ToolResponse.error(e.message ?: "Unknown error"))
        }
    }
    
    @GetMapping("/tools/list")
    fun listTools(): ResponseEntity<List<ToolDefinition>> {
        val tools = listOf(
            ToolDefinition(
                name = "lookup_share_price",
                description = "This tool provides the current price of the given stock symbol.",
                parameters = mapOf(
                    "symbol" to ParameterDefinition("string", "the symbol of the stock", true)
                )
            )
        )
        return ResponseEntity.ok(tools)
    }
}

// Request DTOs
data class LookupSharePriceRequest(val symbol: String)

// Tool definition for MCP discovery
data class ToolDefinition(
    val name: String,
    val description: String,
    val parameters: Map<String, ParameterDefinition>
)

data class ParameterDefinition(
    val type: String,
    val description: String,
    val required: Boolean
)